File: leetcode/2024_6_june/c++/08_continuous_subarray_sum.cpp
// TIME COMPLEXITY
// O(n)

// SPACE COMPLEXITY
// O(n)

// CODE

class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_map<int,int> mp;
                  //<rem , index>
        mp[0] = -1;
        int sum = 0;
        for(int i  = 0; i < n; i++){
            sum += nums[i];
            int rem = sum % k; // calculating remainder
            // map me check kro ki h yaa nhi
            if(mp.find(rem) != mp.end()){
                if(i - mp[rem] >= 2) // check if size is greater or equal than 2
                    return true;
            }
            // nhi ha toh daalo yrr map me
            else{
                mp[rem] = i;
            }
        }
        return false;
    }
};

File: leetcode/2024_6_june/python/08_continuous_subarray_sum.py
# CODE

from typing import List
class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        n = len(nums)
        mapboi = {0 : -1}
        prefix = 0

        for i in range(n):
            prefix = (prefix + nums[i]) % k
            if prefix in mapboi:
                if i - mapboi[prefix] > 1:
                    return True
            else:
                mapboi[prefix] = i
        
        return False